File: evaluation/S2/cpp/Hard/2798-maximum-difference-between-even-and-odd-frequency-ii.cpp
// Source: https://leetcode.com/problems/maximum-difference-between-even-and-odd-frequency-ii/   |   Difficulty: Hard
//
// Problem Description:
// You are given a string s and an integer k. Your task is to find the maximum difference between the frequency of two characters, freq[a] - freq[b], in a substring subs of s, such that:
//
//
// 	subs has a size of at least k.
// 	Character a has an odd frequency in subs.
// 	Character b has a non-zero even frequency in subs.
//
//
// Return the maximum difference.
//
// Note that subs can contain more than 2 distinct characters.
//
// Example:
// Input: s = "12233", k = 4
//
// Output: -1
//
// Explanation:
//
// For the substring "12233", the frequency of '1' is 1 and the frequency of '3' is 2. The difference is 1 - 2 = -1.
//
// Constraints:
// 3 <= s.length <= 3 * 104
// 	s consists only of digits '0' to '4'.
// 	The input is generated that at least one substring has a character with an even frequency and a character with an odd frequency.
// 	1 <= k <= s.length
//
// Helpful references (internal KB):
// - The Inclusion-Exclusion Principle (string, counting)
//   • When to use: Use when you need to determine the frequency of characters or substrings within a string. It is essential for tasks like anagram detection or checking character distribution.
//   • Idea: This technique involves iterating through a string to tally the occurrences of its constituent elements, typically characters or words. It generally operates in O(N) time complexity, where N is the string's length.
//   • Invariants: The count for each character reflects its total occurrences processed so far.; All characters encountered up to the current position have been tallied.
//   • Tips: Use a hash map (dictionary) for arbitrary character sets.; Use an array for fixed-size character sets (e.g., ASCII, lowercase English).
//   • Pitfalls: Forgetting to handle case sensitivity or ignore non-alphanumeric characters.; Incorrectly handling Unicode characters with fixed-size arrays.
// -  
// Kth order statistic in O(N) (array, enumeration)
//   • When to use: Use when you need to find the Kth smallest or largest element in an unsorted array efficiently, without sorting the entire dataset. It's ideal for problems requiring a specific order statistic in linear average time.
//   • Idea: This algorithm finds the Kth smallest (or largest) element in an unsorted array in expected linear time, O(N), by repeatedly partitioning the array around a pivot. It is a selection algorithm, conceptually similar to quicksort but only recursing on one side.
//   • Invariants: All elements to the left of the chosen pivot are less than or equal to the pivot.; All elements to the right of the chosen pivot are greater than or equal to the pivot.
//   • Tips: Choose a good pivot (e.g., random element) to minimize the chance of worst-case performance.; The algorithm can be implemented iteratively to avoid potential recursion depth limits for very large inputs.
//   • Pitfalls: Poor pivot selection can lead to a worst-case time complexity of O(N^2).; Off-by-one errors are common when translating K (1-based) to array indices (0-based).
// - Search the subarray with the maximum/minimum sum (array, prefix-sum, max-window)
//   • When to use: When needing to find the maximum or minimum sum of any contiguous subarray within a given array efficiently. It is particularly useful for problems requiring an O(N) solution.
//   • Idea: This algorithm finds the maximum (or minimum) sum of a contiguous subarray by iterating through the array, maintaining a running prefix sum, and tracking the minimum (or maximum) prefix sum encountered so far. It achieves an optimal O(N) time complexity with O(1) auxiliary space.
//   • Invariants: max_global_sum stores the maximum subarray sum found among all subarrays ending at or before the current element.; min_prefix_sum_value correctly tracks the minimum prefix sum encountered up to the previous element's prefix sum.
//   • Tips: Use a running sum variable instead of explicitly storing the entire prefix sum array.; Keep track of the minimum prefix sum encountered up to the current point.
//   • Pitfalls: Incorrectly initializing the minimum prefix sum (e.g., to INT_MAX instead of 0 or s[0]).; Off-by-one errors when calculating subarray sums using prefix sums (s[r] - s[l-1]).
//
// Notes:
// - Keep the given function signature unchanged.
// - Use standard C++ only; do NOT rely on non-standard extensions or platform-specific APIs.
// - Do not change input/output formats or return value semantics.
// - Conformance: Prefer modern standard C++ (C++11/14/17) features that are portable; avoid compiler-specific extensions (e.g., __int128 unless necessary and supported).
// - If you introduce any helper functions, include their full implementations in this file; no hidden helpers.

class Solution {
public:
    int maxDifference(string s, int k) {
        int n = s.size();
        vector<int> freq(5, 0);
        int oddCount = 0, evenCount = 0;

        // Initialize frequency for the first k characters
        for (int i = 0; i < k; ++i) {
            int digit = s[i] - '0';
            freq[digit]++;
            if (freq[digit] % 2 == 1) {
                oddCount++;
                evenCount -= (freq[digit] > 1) ? 1 : 0;
            } else {
                evenCount++;
                oddCount -= (freq[digit] > 0) ? 1 : 0;
            }
        }

        int maxDiff = (oddCount > 0 && evenCount > 0) ? oddCount - evenCount : INT_MIN;

        // Slide through the string
        for (int i = k; i < n; ++i) {
            // Remove the leftmost character
            int leftDigit = s[i - k] - '0';
            freq[leftDigit]--;
            if (freq[leftDigit] % 2 == 1) {
                oddCount++;
                evenCount -= (freq[leftDigit] >= 0) ? 1 : 0;
            } else {
                evenCount++;
                oddCount -= (freq[leftDigit] >= 0) ? 1 : 0;
            }

            // Add the new character
            int rightDigit = s[i] - '0';
            freq[rightDigit]++;
            if (freq[rightDigit] % 2 == 1) {
                oddCount++;
                evenCount -= (freq[rightDigit] > 1) ? 1 : 0;
            } else {
                evenCount++;
                oddCount -= (freq[rightDigit] > 0) ? 1 : 0;
            }

            // Update max difference
            if (oddCount > 0 && evenCount > 0) {
                maxDiff = max(maxDiff, oddCount - evenCount);
            }
        }

        return maxDiff == INT_MIN ? -1 : maxDiff;           
    }
};

File: evaluation/S0/java/Hard/2798-maximum-difference-between-even-and-odd-frequency-ii.java
// Source: https://leetcode.com/problems/maximum-difference-between-even-and-odd-frequency-ii/   |   Difficulty: Hard
//
// Problem Description:
// You are given a string s and an integer k. Your task is to find the maximum difference between the frequency of two characters, freq[a] - freq[b], in a substring subs of s, such that:
//
//
// 	subs has a size of at least k.
// 	Character a has an odd frequency in subs.
// 	Character b has a non-zero even frequency in subs.
//
//
// Return the maximum difference.
//
// Note that subs can contain more than 2 distinct characters.
//
// Example:
// Input: s = "12233", k = 4
//
// Output: -1
//
// Explanation:
//
// For the substring "12233", the frequency of '1' is 1 and the frequency of '3' is 2. The difference is 1 - 2 = -1.
//
// Constraints:
// 3 <= s.length <= 3 * 104
// 	s consists only of digits '0' to '4'.
// 	The input is generated that at least one substring has a character with an even frequency and a character with an odd frequency.
// 	1 <= k <= s.length
//

class Solution {
    public int maxDifference(String s, int k) {
        int n = s.length();
        int[][] freq = new int[5][n + 1];

        for (int i = 0; i < n; i++) {
            int digit = s.charAt(i) - '0';
            for (int d = 0; d < 5; d++) {
                freq[d][i + 1] = freq[d][i] + (d == digit ? 1 : 0);
            }
        }

        int maxDiff = -1;

        for (int start = 0; start <= n - k; start++) {
            for (int end = start + k; end <= n; end++) {
                for (int a = 0; a < 5; a++) {
                    int freqA = freq[a][end] - freq[a][start];
                    if (freqA % 2 == 1) {
                        for (int b = 0; b < 5; b++) {
                            if (a != b) {
                                int freqB = freq[b][end] - freq[b][start];
                                if (freqB > 0 && freqB % 2 == 0) {
                                    maxDiff = Math.max(maxDiff, freqA - freqB);
                                }
                            }
                        }
                    }
                }
            }
        }

        return maxDiff;         
    }
}
